lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -597,3 +601,18 @@ liftDelegationAddress ix dTemplate (KeyFingerprint fingerprint) =
     delegationCredential = DelegationFromScript . toScriptHash
     dScript =
         replaceCosignersWithVerKeys CA.Stake dTemplate ix
+
+toSharedWalletId
+    :: (WalletKey k, k ~ SharedKey)
+    => k 'AccountK XPub
+    -> ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Digest Blake2b_160
+toSharedWalletId accXPub pTemplate dTemplateM =
+    hash $
+    (unXPub . getRawKey $ accXPub) <>
+    toByteString pTemplate <>
+    maybe mempty toByteString dTemplateM
```

Comment:
maybe we need to put here not `mempty` but `cosigner#0` as we expect in that case multisig wallet to just delegate using `accXPubIx/2/0` where accXPubIx stems from ix of multisig wallet...

Reply:
yes, it needs to be changed. otherwise the following two wallets will give different wallet id, although they are the same:
1. accXPub, spending: something
2. accXPub, spending: something, staking: cosigner#0

Reply:
and I adopted already this approach in https://github.com/input-output-hk/cardano-addresses/pull/185 which should be consistent with what cardano-wallet gives

Reply:
> yes, it needs to be changed. otherwise the following two wallets will give different wallet id, although they are the same:
>
> accXPub, spending: something
> accXPub, spending: something, staking: cosigner#0

Asking as a someone pretty unfamiliar with multi-sig: does this mean you could create "Multi-sig wallet 1" using account 0 and `anyOf [self, Alice]` as script, and then "Multi-sig wallet 2" using account 0 and `anyOf [self, Alice, Bob]` and have those be completely separate wallets despite using the same account for payment keys?

Reply:
> yes, it needs to be changed. otherwise the following two wallets will give different wallet id, although they are the same:
>
>    accXPub, spending: something
>    accXPub, spending: something, staking: cosigner#0

Are you sure, Pawel? 🤔 I suppose that this depends on what exactly the meaning of `Nothing` for `dTemplateM` is. It's perfectly conceivable that these cases are actually:

1. accXPub, spending: something, staking: none (i.e. enterprise address)
2. accXPub, spending: something, staking: cosigner#0

In other words, if there is no script template for delegation, then the wallet only contains enterprise addresses / addresses that do not have a staking part.

Reply:
ok, you are right @HeinrichApfelmus . If we want enterprise addresses to be supported in multisig then we need to allow empty delegation !!!
I will revert commit.
@Anviking  Indeed, same accXPub and different script is different entity, having separate address space, and should be identified separately IMHO

Reply:
Yup. Just for reference: The multi-signature `SharedKey` does indeed create enterprise addresses when the staking template is empty:

```hs
Address $ case dTemplate of
    Just dTemplate' ->
        createBaseAddress pScript (dScript dTemplate')
    Nothing ->
        createEnterpriseAddress pScript
```